### ISSUE TITLE
Using `+` in custom residual Keras `Layer` does not create correct model graph

### ISSUE THREAD

**Issue by jacoblubecki** [tensorflow/tensorflow#54436]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS 10.15.6
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): v2.8.0-rc1-32-g3f878cff5b6 2.8.0
- Python version: 3.7.12

Also reproducible on various hosted Jupyter environments (Kaggle, Colab) with and without GPU.

**Describe the current behavior**

In custom residual block implementation with keras APIs, `+` yields broken graph.

```python
def call(self, x, training=False):
    for block in self.blocks:
        h = x
        for conv in block:
            h = conv(h, training=training)

        x = x + h

    return x
```

This is the resulting graph:
![res-block-broken](https://user-images.githubusercontent.com/7884451/154602227-346a939a-275f-4508-84ff-8b0163e5ac13.png)

**Describe the expected behavior**

This code produces the correct graph, where each `add` is a separate instance of `keras.layers.Add`.
The `+` operator should produce the same graph.

```python
def call(self, x, training=False):
    for block, add in zip(self.blocks, self.adds):
        h = x
        for conv in block:
            h = conv(h, training=training)

        x = add([x, h])

    return x
```
![res-block-working](https://user-images.githubusercontent.com/7884451/154602398-cd61f319-cfee-4bcc-89a3-b1b50e49109d.png)

I will add that this isn't just a visualization issue. My model would not train until after I identified this problem and applied the fixed implementation described above. This was causing serious issues with my gradients and the model could not learn because it was too deep without the skip connections.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): I would consider it, but leaning towards no.
- Briefly describe your candidate solution(if contributing): N/A

**Standalone code to reproduce the issue**
```python
from typing import Optional, Tuple, Union

import tensorflow as tf
import tensorflow.nn
import tensorflow.keras.backend as K
import tensorflow.keras.layers as L
from tensorflow import keras

def plot_model(model, shape):
    inputs = keras.Input(shape[1:])
    ones = tf.ones(shape)
    model(ones)  # I think needed to properly init graph for plotting
    outputs = model.call(inputs)
    wrapped_model = keras.Model(inputs, outputs)
    return tensorflow.keras.utils.plot_model(
        wrapped_model, expand_nested=True, show_shapes=True)

class ConvBnAct(L.Layer):

    def __init__(
        self,
        out_channels: int,
        kernel_size: Union[int, Tuple[int]],
        stride: Union[int, Tuple[int]],
        activation: Optional[str] = 'swish',
        use_bias=False,
        use_batch_norm=True,
        data_format='channels_last'
            ):
        super().__init__()

        self.out_channels = out_channels
        self.kernel_size = kernel_size
        self.stride = stride
        self.use_bias = use_bias
        self.data_format = data_format

        self.activation = L.Activation(activation)
        self.act_type = activation

        bn_axis = 1 if data_format == 'channels_first' else -1
        self.batch_norm = L.BatchNormalization(
            axis=bn_axis) if use_batch_norm else None

    def build(self, input_shape):
        self.conv = L.Conv2D(
            self.out_channels,
            self.kernel_size,
            input_shape=input_shape[1:],
            padding='same',
            strides=self.stride,
            activation=None,
            use_bias=self.use_bias,
            data_format=self.data_format,
            )

    def call(self, inputs, training=False):
        x = self.conv(inputs)

        if self.batch_norm:
            x = self.batch_norm(x, training=training)

        if self.activation:
            x = self.activation(x)

        return x

class ResBlock(L.Layer):

    def __init__(
        self,
        blocks: int,
        shortcut=True,
        data_format='channels_last'
    ):
        super().__init__()
        self.n_blocks = blocks
        self.shortcut = shortcut
        self.data_format = data_format

    def build(self, input_shape):
        channel_axis = 1 if self.data_format == 'channels_first' else -1
        channels = input_shape[channel_axis]

        self.blocks = []
        for i in range(self.n_blocks):
            block = [
                ConvBnAct(channels, kernel_size=1, stride=1, data_format=self.data_format),
                ConvBnAct(channels, kernel_size=3, stride=1, data_format=self.data_format)
                ]
            self.blocks.append(block)

    def call(self, x, training=False):
        for block in self.blocks:
            h = x
            for conv in block:
                h = conv(h, training=training)

            x = x + h if self.shortcut else h

        return x

if __name__ == '__main__':
    i = keras.Input((24, 24, 3))
    r = ResBlock(2, True)
    plot_model(r, (1, 24, 24, 3))
```

**Comment by Cheril311** [comment]
@jacoblubecki I guess you cannot add layers using the '+' operator, as I think it would be confusing if that was allowed

**Comment by jacoblubecki** [comment]
@Cheril311 Sorry if it was a bit unclear -- the `+` operator is being used to add the output tensors of the layers (_NOT_ adding the layers themselves).

Both versions of the code work during the forward pass. When setting random seeds and `TF_DETERMINISTIC_OPS=1`, _**the outputs of each version are identical**_.

Problems only appear during training because the gradients are not computed correctly for the `+`-version due to the malformed computation graph, and the training problems are entirely non-obvious because the backwards pass executes and the model's weights do get updated, but the model fails to converge because the gradients aren't correct.

**Comment by sushreebarsa** [comment]
@jacoblubecki 
Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see;
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)
Thank you!

**Comment by jacoblubecki** [comment]
@sushreebarsa I have filed a ticket here:  https://github.com/keras-team/keras/issues/16098

I replaced my `ConvBnAct` with the following:
```python
class ConvBnAct(object):

    def __init__(
        self,
        out_channels: int,
        kernel_size: Union[int, Tuple[int]],
        stride: Union[int, Tuple[int]],
        activation: Optional[str] = 'swish',
        use_bias=False,
        use_batch_norm=True,
        data_format='channels_last'
            ):
        super().__init__()
        self.kernel = tf.ones((kernel_size, kernel_size, 3, 3))

    def __call__(self, inputs, training=False):
        x = tf.nn.conv2d(inputs, self.kernel, strides=1, padding='SAME')
        return x
```

Then I used tensorboard to trace two versions of `ResBlock`.

Keras version using only `tensorflow` APIs (still inherits from Keras `Layer`, but otherwise only using `tf` namespace):
```python
class ResBlock(L.Layer):

    def __init__(self, blocks: int, data_format='channels_last'):
        super().__init__()
        self.n_blocks = blocks

        self.blocks = []
        for i in range(self.n_blocks):
            block = [
                ConvBnAct(3, kernel_size=1, stride=1, data_format=self.data_format),
                ConvBnAct(3, kernel_size=3, stride=1, data_format=self.data_format)
                ]
            self.blocks.append(block)

    def call(self, x, training=False):
        for block in self.blocks:
            h = x
            for conv in block:
                h = conv(h, training=training)

            x = x + h

        return x
```
And pure tensorflow version (with no reference to any Keras APIs):
```python
class ResBlock(object):

    def __init__(self, blocks: int, data_format='channels_last'):
        super().__init__()
        self.n_blocks = blocks

        self.blocks = []
        for i in range(self.n_blocks):
            block = [
                ConvBnAct(3, kernel_size=1, stride=1, data_format=data_format),
                ConvBnAct(3, kernel_size=3, stride=1, data_format=data_format)
                ]
            self.blocks.append(block)

    def __call__(self, x, training=False):
        for block in self.blocks:
            h = x
            for conv in block:
                h = conv(h, training=training)

            x = x + h

        return x
```

The Keras version is still broken on tensorboard, but the pure version with no Keras APIs appears to work fine.

**Comment by sushreebarsa** [comment]
@jacoblubecki Could you please let us know if we can move this issue to closed status as we are addressing the other ticket in  [keras-team/keras repo.](https://github.com/keras-team/keras/issues) ?
Thanks!

**Comment by jacoblubecki** [comment]
@sushreebarsa Sure thing, thanks for the redirect.
